### PR TITLE
14 - distributed computation, ft AsyncIO, Spark, and Dask

### PR DESCRIPTION
**Pull request by woozyking**
preview: https://github.com/EQWorks/python-curriculum/blob/14/data-analysis/14-distributed-computation/notebook.ipynb

### CODE REVIEW

**Review by maluhoss — COMMENTED** [review]
looks great! learned quite a bit too in the process. I like how you included the Spark UI screenshots + went step by step explaining the time diffs and tradeoffs as well in using Spark.

Some typos and suggestions:
- suggestion: maybe include the image of multiprocessing vs multithreading in terms of memory space - there was a diagram from our part from what I remember
- 'less computing resources compared to its OS-native sibling' - this is threads right?
- 'No explicit event loop initiation (it's already in one)' - maybe mention that this is how each of these cell blocks run in order
- typo: 'Let's use the responses to verify the ELAPSED time of the API calls by using functional techniques map() and reduce()'
- maybe have an image of this mapreduce model - or if you can't find a good one, I can help create one for you
- typos: 'Spark starts the computation, along with other PROCESSES once we INSTRUCT it to take the first 10 values from the RDD. If we go back to the Spark UI, it may paint a better picture on what Spark has done under the hood.'
- typo: 'Spark DataFrames can also be created directly, or through the help of CONCATENATING Pandas DataFrames.'
- typo: 'The above examples demonstrate the SAMPLING process may be more time consuming'
- missing word: 'After A certain threshold, it exhibits a linear growth in the time needed to perform the operation'
- typo: 'At EQ Works, we leverage a cloud provider managed Spark service to run through petabytes of data with variance of computational complexities in hours or minutes which would otherwise take days or even months on a single machine, even if the said machine can hold LARGE amounts of data reliably.'
- suggestion: have this EQ fact above when you first introduce Spark. I think it draws people in from the get-go to hear this fact :) 

There was another typo I remember seeing but now I can't seem to find it (or mind is playing tricks on me) but if you find it, great! if not, no worries.

**Comment by woozyking** [comment]
> looks great! learned quite a bit too in the process. I like how you included the Spark UI screenshots + went step by step explaining the time diffs and tradeoffs as well in using Spark.
> 
> Some typos and suggestions:

I plan to run this through Grammarly after integrating additional examples for typo and grammar mistakes so will group them together, will focus on the other suggestions

> 
>     * suggestion: maybe include the image of multiprocessing vs multithreading in terms of memory space - there was a diagram from our part from what I remember

Yes, will make link-backs in the material itself, and during the session have part 13 ready on the side to jump back to

> 
>     * 'less computing resources compared to its OS-native sibling' - this is threads right?

Yes, maybe the context can be a bit more pronounced, I'll group this together with typo/grammar grooming later.

> 
>     * 'No explicit event loop initiation (it's already in one)' - maybe mention that this is how each of these cell blocks run in order

Yes, this will be mentioned during the session, but less worried about the material itself as the knowledge should be notebook-independent. However, I can try to clarify a bit more (maybe carry the 2 tips inside the code block out in its dedicated markdown text block).

>     * maybe have an image of this mapreduce model - or if you can't find a good one, I can help create one for you

That'll be great @maluhoss , the ones I have found so far are not as illustrative, maybe only https://www.tutorialspoint.com/map_reduce/map_reduce_introduction.htm has something that's simple enough to digest

>     * missing word: 'After A certain threshold, it exhibits a linear growth in the time needed to perform the operation'
👍 

>     * suggestion: have this EQ fact above when you first introduce Spark. I think it draws people in from the get-go to hear this fact :)

Agree, and will have this incorporated at least during the session.

> 
> There was another typo I remember seeing but now I can't seem to find it (or mind is playing tricks on me) but if you find it, great! if not, no worries.

👍 I'm sure after Grammarly grooming there will be a bit less typo noise. So tomorrow we can give it another round of check

File: README.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ _WIP_
 * [Part 11 - Work with SQL](data-analysis/11-work-with-sql.ipynb)
 * [Part 12 - Generate data](data-analysis/12-generate-data.ipynb)
 * [Part 13 - The Power of Parallel Processing](data-analysis/13-data-processing.ipynb)
+* [Part 14 - Distributed Computation](data-analysis/14-distributed-computation/notebook.ipynb)
```
Comment (divyselva) [inline README.md:35]: MapReduceImage

Comment (divyselva) [reply README.md:35]: Picture for explaining MapReduce ?

Comment (woozyking) [reply README.md:35]: looks good!

**Review by maluhoss — APPROVED** [review]
looks good! good choice for mapreduce diagram. Just noticed some formatting issue here:

Spark would attempt optimization not to scan the full range of dataset when invoking partial readings such as `rdd.take()` or `df.show()`, <-- like it shows the backticks in the markdown. Visual is def good to have for a concept like this though

Also the RDD diagram looks a little complicated. The notes made sense to me on what it was but the diagram confused me a bit. Assuming you'll go over it during the session but maybe consider having like one line that gives context to the diagram

**Comment by woozyking** [comment]
> Spark would attempt optimization not to scan the full range of dataset when invoking partial readings such as `rdd.take()` or `df.show()`, <-- like it shows the backticks in the markdown. Visual is def good to have for a concept like this though

ah yes, you know the trend/standard development of having punctuations go before the closing quotes? Grammarly probably messed it up automatically lol will fix

> 
> Also the RDD diagram looks a little complicated. The notes made sense to me on what it was but the diagram confused me a bit. Assuming you'll go over it during the session but maybe consider having like one line that gives context to the diagram

will have to articulate for sure

**Pull request merged**